src/common/modules/escrow/escrow.service.ts

```diff
@@ -43,17 +58,42 @@ export class EscrowService {
   }
 
   async orderFulfilled(order): Promise<void> {
-    try {
-      const tokenContract = this.ethereumService.getEscrowSmartContract();
-      const wallet: WalletSigner = await this.ethereumService.createWallet(
-        this.process.env.DEBIO_ESCROW_PRIVATE_KEY,
-      );
-      const tokenContractWithSigner = tokenContract.connect(wallet);
-      const tx = await tokenContractWithSigner.fulfillOrder(order.id);
-      console.log('fullfilled order customerId :', order.customerId, ' ->', tx);
-    } catch (error) {
-      console.log(error);
-    }
+    let currentNonce;
+    lock
+      .acquire(ESCROW_WALLET_LOCK_KEY, async () => {
+        const _nonce = await this.provider.getTransactionCount(
+          this.escrowWallet.address,
+        );
+        nonce = nonce > _nonce ? nonce : _nonce;
+        const feeData = await this.provider.getFeeData();
+        const gasPrice = feeData.gasPrice;
+        const tokenContract = this.ethereumService.getEscrowSmartContract();
+        const tokenContractWithSigner = tokenContract.connect(
+          this.escrowWallet,
+        );
+        const tx = await tokenContractWithSigner.fulfillOrder(order.id, {
+          nonce,
+          gasPrice,
+        });
+        currentNonce = nonce;
+        this.provider.waitForTransaction(tx.hash).then((_tx) => {
+          console.log(
+            'fullfilled order customerId :',
+            order.customerId,
+            ' ->',
+            _tx,
+          );
+        });
+        nonce += 1;
+      })
+      .then(function () {
+        console.log(
+          `[fulfillOrder] Sent transaction for nonce: ${currentNonce}`,
+        );
+      })
+      .catch(function (err) {
+        console.log(err);
+      });
```

Comment:
What is this for?

Reply:
this one is for locking mechanism, so that the nonce will not be duplicated. got the reference from https://betterprogramming.pub/sending-web3-transactions-in-node-js-nonce-hell-f3ba82edbf3d. have tried without locking but still got an error. it's correlated with single threaded nodejs.

Reply:
Why does the `ESCROW_WALLET_LOCK_KEY` have to be out of scope?

Why is the `nonce` out of scope?

Is there a way for you to simplify this?

Reply:
@JackyRahman @hilyds

Reply:
Have you tried executing without the `lock`?

Reply:
@agustinustheo because the `nonce` and `ESCROW_WALLET_LOCK_KEY` will later be used in other function such as refundOrder

Reply:
You don't use `nonce` or `ESCROW_WALLET_LOCK_KEY` in `refundOrder`. Please add the code first and then ask me to re-review.